src/connector/src/sink/redis.rs

```diff
@@ -40,28 +108,302 @@ impl Sink for RedisSink {
 
     const SINK_NAME: &'static str = "redis";
 
-    async fn new_log_sinker(&self, _writer_env: SinkWriterParam) -> Result<Self::LogSinker> {
-        todo!()
+    async fn new_log_sinker(&self, writer_param: SinkWriterParam) -> Result<Self::LogSinker> {
+        Ok(RedisSinkWriter::new(
+            self.config.clone(),
+            self.schema.clone(),
+            self.pk_indices.clone(),
+            self.is_append_only,
+        )?
+        .into_log_sinker(writer_param.sink_metrics))
     }
 
     async fn validate(&self) -> Result<()> {
-        todo!()
+        let client = self.config.common.build_client()?;
+        client.get_connection()?;
+        let all_set: HashSet<String> = self
+            .schema
+            .fields()
+            .iter()
+            .map(|f| f.name.clone())
+            .collect();
+        let pk_set: HashSet<String> = self
+            .schema
+            .fields()
+            .iter()
+            .enumerate()
+            .filter(|(k, _)| self.pk_indices.contains(k))
+            .map(|(_, v)| v.name.clone())
+            .collect();
+        match (
+            &self.config.common.key_format,
+            &self.config.common.value_format,
+        ) {
+            (Some(key_format), Some(value_format)) => {
+                check_string_format(key_format, &pk_set)?;
+                check_string_format(value_format, &all_set)?;
+            }
+            (Some(key_format), None) => {
+                check_string_format(key_format, &pk_set)?;
+            }
+            (None, Some(value_format)) => {
+                check_string_format(value_format, &all_set)?;
+            }
+            (None, None) => {}
+        };
+        Ok(())
+    }
+}
+
+pub struct RedisSinkWriter {
+    epoch: u64,
+    schema: Schema,
+    is_append_only: bool,
+    pk_indices: Vec<usize>,
+    formatter: SinkFormatterImpl,
+    payload_writer: RedisSinkPayloadWriter,
+}
+
+struct RedisSinkPayloadWriter {
+    // connection to redis, one per executor
+    conn: Option<Connection>,
+    // the command pipeline for write-commit
+    pipe: Pipeline,
+}
+impl RedisSinkPayloadWriter {
+    pub fn new(config: RedisConfig) -> Result<Self> {
+        let client = config.common.build_client()?;
+        let conn = Some(client.get_connection()?);
+        let pipe = redis::pipe();
+        Ok(Self { conn, pipe })
+    }
+
+    #[cfg(test)]
+    pub fn mock() -> Self {
+        let conn = None;
+        let pipe = redis::pipe();
+        Self { conn, pipe }
+    }
+
+    pub fn commit(&mut self) -> Result<()> {
+        self.pipe.query(&mut self.conn.as_mut().unwrap())?;
+        self.pipe.clear();
+        Ok(())
     }
 }
 
-pub struct RedisSinkWriter;
+impl FormattedSink for RedisSinkPayloadWriter {
+    type K = String;
+    type V = Vec<u8>;
```

Comment:
IIRC, `ToRedisArg` accepts bytes for both key and value. any reason for limiting key to string?

Reply:
Inserting `k v` into redis in string format is only supported here.